llvm/include/llvm/ProfileData/ProfileFuncRef.h

```diff
@@ -0,0 +1,222 @@
+  static int compareMemory(const char *Lhs, const char *Rhs, uint64_t Length) {
+    if (Lhs == Rhs)
+      return 0;
+    if (!Lhs)
+      return -1;
+    if (!Rhs)
+      return 1;
+    return ::memcmp(Lhs, Rhs, (size_t)Length);
+  }
+
+public:
+  ProfileFuncRef() = default;
+
+  /// Constructor from a StringRef. Check if Str is a number, which is generated
+  /// by converting a MD5 sample profile to a format that does not support MD5,
+  /// and if so, convert the numerical string to a hash code first. We assume
+  /// that no function name (from a profile) can be a pure number.
+  explicit ProfileFuncRef(StringRef Str)
+      : Data(Str.data()), LengthOrHashCode(Str.size()) {
+    // Only need to check for base 10 digits, fail faster if otherwise.
+    if (Str.size() > 0 && isdigit(Str[0]) &&
+        !Str.getAsInteger(10, LengthOrHashCode))
+      Data = nullptr;
+  }
+
+  /// Constructor from a hash code.
+  explicit ProfileFuncRef(uint64_t HashCode)
+      : Data(nullptr), LengthOrHashCode(HashCode) {
+    assert(HashCode != 0);
+  }
+
+  /// Check for equality. Similar to StringRef::equals, but will also cover for
+  /// the case where one or both are hash codes. Comparing their int values are
+  /// sufficient. A hash code ProfileFuncRef is considered not equal to a
+  /// StringRef ProfileFuncRef regardless of actual contents.
+  bool equals(const ProfileFuncRef &Other) const {
+    return LengthOrHashCode == Other.LengthOrHashCode &&
+           compareMemory(Data, Other.Data, LengthOrHashCode) == 0;
+  }
+
+  /// Total order comparison. If both ProfileFuncRef are StringRef, this is the
+  /// same as StringRef::compare. If one of them is StringRef, it is considered
+  /// greater than the hash code ProfileFuncRef. Otherwise this is the the same
+  /// as comparing their int values.
+  int compare(const ProfileFuncRef &Other) const {
+    auto Res = compareMemory(
+        Data, Other.Data, std::min(LengthOrHashCode, Other.LengthOrHashCode));
+    if (Res != 0)
+      return Res;
+    if (LengthOrHashCode == Other.LengthOrHashCode)
+      return 0;
+    return LengthOrHashCode < Other.LengthOrHashCode ? -1 : 1;
+  }
+
+  /// Convert to a string, usually for output purpose.
+  std::string str() const {
+    if (Data)
+      return std::string(Data, LengthOrHashCode);
+    if (LengthOrHashCode != 0)
+      return std::to_string(LengthOrHashCode);
+    return std::string();
+  }
+
+  /// Convert to StringRef, a backing buffer must be provided in case this
+  /// object represents a hash code, which will be converted to a string into
+  /// the buffer. If this object represents a StringRef, the buffer is not used.
+  StringRef stringRef(std::string &Buffer) const {
+    if (Data)
+      return StringRef(Data, LengthOrHashCode);
+    if (LengthOrHashCode != 0) {
+      Buffer = std::to_string(LengthOrHashCode);
+      return Buffer;
+    }
+    return StringRef();
+  }
```

Comment:
This API feels klunky. Would it be possible to just have a `name()` (potentially returning StringRef() if there is only a hashcode) and a `std::optional<uint64_t> hashcode()` if necessary? This way you can also avoid the buffering logic that may or may not fit the callers needs...

Reply:
If we know for sure MD5 is not used, then this object can just be pointer casted to StringRef, and if we know for sure MD5 is used, then we Have to use a string buffer to support the generated string, and we cannot return a std::string either because an it's a temporary. In the previous implementation we do the opposite in a way no better, that when we create StringRef for MD5 values, we have to maintain a set of strings in the class object as backing buffer, which makes its lifetime much longer than actually needed.